File: packages/engine/src/cpus/6502/instructions/flags.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('clear instructions', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.carry = true
  cpu.interruptDisabled = true
  cpu.overflow = true
  mmu.data = [0x18, 0xd8, 0x58, 0xb8] // CLC CLD CLI CLV
  const g = cpu.run()
  for (let i = 0; i < 8; i++) g.next()
  expect(cpu.carry).toBeFalsy()
  expect(cpu.interruptDisabled).toBeFalsy()
  expect(cpu.overflow).toBeFalsy()
})

test('set instructions', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.carry = false
  cpu.interruptDisabled = false
  mmu.data = [0x38, 0xf8, 0x78] // SEC SED SEI
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.carry).toBeTruthy()
  expect(cpu.interruptDisabled).toBeTruthy()
})

File: packages/engine/src/cores/famibob/cpu/instructions/flags.test.zig
const testing = @import("../testing/setup.zig");

test "clear instructions" {
    var cpu = testing.createCPU();
    cpu.c = true;
    cpu.i = true;
    cpu.v = true;
    cpu.write(0, 0x18); // CLC
    cpu.write(1, 0xd8); // CLD
    cpu.write(2, 0x58); // CLI
    cpu.write(3, 0xb8); // CLV
    for (0..8) |_| {
        cpu.process();
    }
    try testing.expect(!cpu.c and !cpu.i and !cpu.v);
}

test "set instructions" {
    var cpu = testing.createCPU();
    cpu.c = false;
    cpu.i = false;
    cpu.write(0, 0x38); // SEC
    cpu.write(1, 0xf8); // SED
    cpu.write(2, 0x78); // SEI
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expect(cpu.c and cpu.i);
}
